socs/agents/hwp_pid/agent.py

```diff
@@ -27,88 +84,116 @@ def __init__(self, agent, ip, port, verbosity):
+            {
+                "current_freq": current_freq,
+                "target_freq": target_freq,
+                "direction": direction,
+                "last_updated": time.time(),
+            }
+        )
+        self.agent.publish_to_feed("hwppid", data)
+
+    def _process_actions(self, pid):
+        while not self.action_queue.empty():
+            action = self.action_queue.get()
+            try:
+                self.log.info(f"Running action {action}")
+                res = action.process(pid)
+                action.deferred.callback(res)
+            except Exception as e:
+                self.log.error(f"Error processing action: {action}")
+                action.deferred.errback(e)
+
+    def _clear_queue(self):
+        while not self.action_queue.empty():
+            action = self.action_queue.get()
+            action.deferred.errback(Exception("Action cancelled"))
 
+    def main(self, session, params):
+        """main()
+
+        **Process** - Main Process for PID agent. Periodically queries PID
+            controller for data, and executes requested actions.
+
+        Notes:
+            The most recent data collected is stored in the session data in the
+            structure::
+
+                >>> response.session['data']
+                {'current_freq': 0,
+                 'target_freq': 0,
+                 'direction': 1,
+                 'last_updated': 1649085992.719602}
         """
-        if self._initialized and not params['force']:
-            self.log.info("Connection already initialized. Returning...")
-            return True, "Connection already initialized"
+        pid = pd.PID(ip=self.ip, port=self.port, verb=self._verbosity)
+        self.log.info("Connected to PID controller")
 
-        with self.lock.acquire_timeout(10, job='init_connection') as acquired:
-            if not acquired:
-                self.log.warn(
-                    'Could not run init_connection because {} is already running'.format(self.lock.job))
-                return False, 'Could not acquire lock'
+        self._clear_queue()
 
-            try:
-                self.pid = pd.PID(ip=self.ip, port=self.port,
-                                  verb=self._verbosity)
-                self.log.info('Connected to PID controller')
-            except BrokenPipeError:
-                self.log.error('Could not establish connection to PID controller')
-                reactor.callFromThread(reactor.stop)
-                return False, 'Unable to connect to PID controller'
+        sample_period = 5.0
+        last_sample = 0.0
+        session.set_status("running")
+        while session.status in ["starting", "running"]:
+            now = time.time()
+            if now - last_sample > sample_period:
+                self._get_data_and_publish(pid, session)
+                last_sample = now
 
-        self._initialized = True
+            self._process_actions(pid)
+            time.sleep(0.2)
```

Comment:
Compute from `sample_period`?

Also, better to use a pacemaker?

Reply:
`sample_period` is meant to be the rate at which the PID variables are queried and written to feed. This sleep time is for the faster action processing loop, which is shorter than the `samle_period`. The fact that 0.2 sec is `1/5.0` is coincidence. The handling of the sampling period is done in the block above:
```
            if now - last_sample > sample_period:
                self._get_data_and_publish(pid, session)
                last_sample = now
```

I could add the pacemaker here for the .2 sec action loop, however I'm not sure if that is necessary because the daq/data publishing is handled separately (by the code above) and so that wouldn't necessarily regulate the publish rate.